Expose HeightfieldColliderShape to use it in Game Studio

EternalTamago opened the pull request:
Updated 2020/03/31

# PR Details

<!--- Provide a general summary of your changes in the Title above -->
Expose HeightfieldColliderShape to use it in Game Studio.

![scene_heightfield](https://user-images.githubusercontent.com/13661631/69827045-1a3fce80-1259-11ea-8e62-e2b88b72d0d1.png)


* Add HeightmapAsset used to set initial heights of heightfield.
* Add HeightfieldColliderShapeDesc to expose HeightfieldColliderShape in Game Studio.
* Add AlwaysUpdateNaviMeshCache property to StaticColliderComponent to update the cache at rebuilding navigation mesh.

## Description

<!--- Describe your changes in detail -->
### HeightmapAsset
Used to initialize heights of heightfield.

![add_heightmap](https://user-images.githubusercontent.com/13661631/69827108-5b37e300-1259-11ea-9434-c017756ef213.png)

![heightmap_asset_thumbnail](https://user-images.githubusercontent.com/13661631/69827153-7d316580-1259-11ea-8a10-b1a15e92cd94.png)

![heightmap_asset_property](https://user-images.githubusercontent.com/13661631/77956289-9d0e8e00-730c-11ea-9702-079379b13626.png)


#### Source
An image file used to generate heightmap from its R channels.

#### Floating Point Component Range
Determine the range of the floating-point component when the R component is the floating point component.

#### Is Symmetric Short Component
Determine the range of the signed short integer component when the R component is the signed short integer component.
If enabled, [-32767 .. 32767], otherwise [-32768 .. 32767].

#### s RGB sampling
Enable if needed to load the image file as sRGB.

#### Height
Type of height.

##### Height Range
X is min height and Y is max height. "height * HeightScale" should be in this range. In Byte height type, the range can't include both of positive and negative.

##### Height Scale
In Short type or in Byte type, "height * HeightScale" is used for collisions in physics. In Float type, heights are used as they are.

|item|description|
|:---|------|
|Auto|HeightScale Calculated from HeightRange|
|Custom|Custom HeightScale|

#### Resize
New size of the heightmap. X is width and Y is length. If disabled, the size is same to the image size(width * height). Width and length should be greater than or equal to 2.

#### Scale To Height Range
If enabled, scale each of the heights to the height range before they are stored as heightmap. By the default, they are considered to be in [-1 .. 1] when the R component is floating-point component format. Edit FloatingPointComponentRange to actual floating point component range(e.g. [0 .. 1]), if needed.

### HeightfieldColliderShapeDesc

Used to create HeightfieldColliderShape.

![heightfield_inline_property](https://user-images.githubusercontent.com/13661631/77956445-d8a95800-730c-11ea-91a7-fcbb5b3ee522.png)


#### Source
The source to initialize heights.

##### Heightmap
The heightmap to initialize heights.

##### Float, Short and Byte
Type of height.

###### Size
Size of the heightfield. X is width and Y is length. Width and length should be greater than or equal to 2. The Size should be 65 * 65 when you want 64 * 64 size in a scene.

###### Height Range
X is min height and Y is max height. "height * HeightScale" should be in this range. In Byte height type, the range can't include both of positive and negative.

###### Height Scale
In Short type or in Byte type, "height * HeightScale" is used for collisions in physics. In Float type, heights are used as they are.

|item|description|
|:---|------|
|Auto|HeightScale Calculated from HeightRange|
|Custom|Custom HeightScale|

#### Flip Quad Edges
Change how triangles forming heightfield collider shape are generated.(I'm not sure about this)

#### Centering
Add offset to ColliderShape.LocalOffset in order to center specific height.

### StaticColliderComponent.AlwaysUpdateNaviMeshCache
Enable if you want to update the cache of the collider shape at rebuilding the NavigationMesh.

Use one of the following arrays to update the heights at runtime.
|HeightType|Heights of heightfield|
|:---|--|
|Float|HeightfieldColliderShape.FloatArray|
|Short|HeightfieldColliderShape.ShortArray|
|Byte|HeightfieldColliderShape.ByteArray|

## Related Issue

<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
#494

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
Currently there is HeightfieldColliderShape but it can be created only at runtime. It's not available in Game Studio and NavigationMesh.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Docs change / refactoring / dependency upgrade
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

HeadClot commented:
I love this PR.

xen2 commented on sources/engine/Xenko.Physics/Engine/Heightmap.cs:
```diff
@@ -0,0 +1,82 @@
+                    HeightfieldType = HeightfieldTypes.Float,
+                    Width = width,
+                    Length = length,
+                    Floats = new List<float>(data as float[]),
```
Wouldn't it make more sense to make `Heightmap.Floats`, `Heightmap.Shorts` and such directly as arrays, and avoid the two extra copy? (one copy here and one copy in `CreateHeights` because of `ToArray()`)

EternalTamago replied:
Yes, those should be arrays.

xen2 commented on sources/engine/Xenko.Physics/Shapes/HeightfieldColliderShape.cs:
```diff
@@ -48,7 +68,7 @@ bool flipQuadEdges
             HeightStickLength = heightStickLength;
             HeightType = heightType;
-            HeightScale = heightScale;
+            HeightScale = (HeightType == HeightfieldTypes.Float) ? 1f : heightScale;
```
Seems strange/misleading to ignore this parameter only in case of float?

EternalTamago replied:
I revert that.
That is unnecessary because HeightfieldColliderShapeDesc does it.

xen2 commented:
Tried with a heightmap but couldn't make it work properly:

![image](https://user-images.githubusercontent.com/527565/70146281-30f38480-16a2-11ea-81db-7d1afb83392d.png)

I tried float/short/byte (set to same value on both component collider shape's list and heightmap asset).
Anything I might have forget?

Also small remark, maybe `HeightmapAsset.Size` might be better to default to something meaningful like 1024 rather than 0? (even if not enabled)

EternalTamago commented:
It seems that Heightmap size is not same to HeightStickSize.
Does that happen even if those are matched? (Enable Resize of HeightmapAsset and set it to 80x80)
Heightfield uses heightmap as it is, that doesn't try to match the size.

>HeightmapAsset.Size

Yes, I think so too.
It might be better that that is same to default value of HeightStickSize?

EternalTamago commented:
Would it better to change the list of the property to this?

![another_property_type](https://user-images.githubusercontent.com/13661631/70221578-f82fda00-178b-11ea-901b-8e5771842311.png)


BTW, I realized that HeightfieldColliderShape.FloatArray/ShortArray/ByteArray should be locked while building the navigation mesh.

xen2 commented:
@EternalTamago By this change, all the size/type info would come from the collider shape instead of being in two places, right?
If yes, that would be ideal! (I was going to suggest this too so that it is more user-friendly)

EternalTamago commented:
@xen2 I didn't quite understand what "from the collider shape" means.
HeightStickSize and HeightType will be from the Heightmap when the Heightmap is selected at InitialHeights.

I am currently encountering an access violation exception that I do not know the cause of.
The exception occurs in btCollisionWorld_rayTest (called in the PlayerInput script of the TopDownRPG project) after rebuilding the DynamicNavigationMesh.
Or the game will be killed without that exception at similar situation.
It does not always happen.
So far, this only happens when there is a CompoundColliderShape that contains a HeightfieldColliderShape.

EternalTamago commented:
I think that the problem has been fixed.
The cause was that the collider shape was recreated at rebuilding NavigaitonMesh.

xen2 commented:
Works much better for me in new version!

Noticed a few things that either didn't work or that I might have misunderstood:
* `HeightfieldColliderShapeDesc.HeightScale` doesn't work for float
* `HeightfieldColliderShapeDesc.HeightScale` behave strangely for non-float: enabled and not-enabled with size 1 result in different (I think enabled results in 255)
* If I change `HeightfieldColliderShapeDesc.HeightRange` to [0..10], I see everything under the Y axis ground plane (Y < 0) where I would expect it to be between Y=0 and Y=10.

EternalTamago commented:
@xen2 

1. HeightfieldColliderShapeDesc.HeightScale is used for Short or Byte.
When Float, HeightfieldColliderShapeDesc.HeightScale is not used because heightfield uses FloatArray as it is.
**Update**: Changed display name of HeightfieldColliderShapeDesc.HeightScale.

2. When disabled, HeightScale was set internally.
**Update**: Changed type of HeightScale of desc to be better.

3. Recentering happens when the HeightRange is not symmetric.
**Update**: Added an option to add extra offset to the local offset to counteract recentering.

EternalTamago commented:
Sorry to change many times after review request.

I would like to move `HeightfieldColliderShapeDesc.HeightRange` and `HeightfieldColliderShapeDesc.HeightScale` to `HeightmapAsset`.
That might make it more friendly.
What do you think about that?

xen2 commented:
@EternalTamago I think it makes sense UX-wise to move those two properties on the asset itself.

EternalTamago commented:
@xen2 Thanks for the reply.
I will modify this once more.

EternalTamago commented:
@xen2 Sorry for the delay!
I've finished that.

EternalTamago commented:
I have started thinking GetStaticColliderCacheSettings is bad method.

To decide whether to update the navigation mesh cache object at runtime,
it might be better to add something like AlwaysUpdateNaviMeshCache property to StaticColliderComponent instead of adding StaticColliderCacheSettings to StaticColliderData.
However, the property will be useless for the physics and that will be used only at building/rebuilding NavigationMesh.

EternalTamago commented:
@xen2 I'm sorry to bother you again.

I will try to add AlwaysUpdateNaviMeshCache to StaticColliderComponent in order to decide wthether to update the cache at rebuilding NavigationMesh.
And GetStaticColliderCacheSettings will be removed.

Also I'd like to update parts of HeightfieldColliderShapeDesc in order to be easy to understand and to be able to set initial height value that might be needed when the height range is not include 0.

xen2 commented:
@EternalTamago Hey, in few days Xenko will be renamed to Stride so it's easier to merge PRs before that. Let me know if you have a working version that can be merged already.
Otherwise no worries, I will guide you on how to make sure PR don't conflict with the renaming whenever you are ready.

EternalTamago commented:
@xen2 Yes. Sorry to late response!

EternalTamago commented:
@xen2 The current state is everything. My words were not enough. Sorry if you are blocked by me.

xen2 commented:
@EternalTamago 
Thanks for clarification! No worries, I understood it was complete with previous messages.
I wanted to test it quickly before merging, sorry for the delay.

Aggror commented:
@EternalTamago Hi EternalTamago,
You wouldn't happen to have a nice video recording that demonstrates the heightmap field with a nice example scene? I want to compile a nice video containing the cool new features for the Stride 4.0 release. 
If you have anything resource wise that can help out with showcasing your heightfielf PR, you can contact me on Discord #Aggror#6358 or here on Gitbhub.

EternalTamago commented:
Hi, @Aggror 
I'll try to record simple video with Stride sample assets etc in a few days.
(You don't have to wait for it when we don't have the time.)

Kryptos-FR commented:
@EternalTamago @Aggror 
Digging some old tweets 😉 :
https://twitter.com/EternalTamago/status/1045021643620274176
https://twitter.com/EternalTamago/status/1045742093786787840

Aggror commented:
Note that we will now just use small videos in the release notes instead of one long video.

EternalTamago commented:
@Aggror 
I have sent the video to you via DM in Discord.

@Kryptos-FR 
New video may be same as the above mostly except that bitrate 😄

EternalTamago commented:
@xen2 
Recently I posted separately #734 from #494.
And I'd like to know what you think about #734 and the heightfield.
That is related to the stability of Stride.

For me, the heightfield collider shape in Game Studio is useful even if there is that issue.
However, the issue might be critical for some users that try to use the heightfield.
What do you think about this?

I know that it's too late to talk, but I think that we should get clarity on this before the release.
I'm sorry to bother you.
